### PR TITLE
Experimental prototype of Swift quasiquotes

### PR DESCRIPTION
**Pull request by burmako**
Swift works great as an infinitely hackable syntactic interface to
semantics that are defined by the compiler underneath it.
The two options today are LLVM (there's a running joke that Swift is
just syntactic sugar for LLVM) and TensorFlow graphs (which is the
contribution of early versions of Swift for TensorFlow).

Multi-Level Intermediate Representation (MLIR) is a generalization of
both the LLVM IR and TensorFlow graphs to represent arbitrary
computations at multiple levels of abstraction to enable
domain-specific optimizations and code generation (e.g. for CPUs, GPUs,
TPUs, and other hardware targets).

In https://docs.google.com/document/d/1UIPWl4lvBTozBD5OQ9SrxgcM7rA4pODMOjqQv3tm57w/edit,
we've written down some thoughts on several ways to metaprogram MLIR
in Swift - starting from treating MLIR programs as strings and then
gradually increasing the level of language integration with Swift.

Seeking to obtain experimental evaluation of the designs explored
in the document, in this commit, we're introducing an experimental
prototype of quasiquotes, a language feature that allows "quoting"
snippets of code which are then transformed into data structures
available to Swift programs. These data structures can then be used
for all sorts of purposes, including translation to MLIR.

This code doesn't fully implement the theorized design yet, but we
believe that it is already useful for experimentation. We are evaluating
available approaches and garnering community feedback. Please let us
know what you think!

### CODE REVIEW

**Comment by burmako** [comment]
Here's the library portion of this pull request: https://github.com/tensorflow/swift/pull/225. Check it out for detailed documentation of supported language constructs and some examples of use

File: utils/gyb_syntax_support/ExprNodes.py [hunk 1]
```diff
@@ -462,7 +462,7 @@
          traits=['Parenthesized'],
          children=[
              Child('Backslash', kind='BackslashToken'),
-             Child('Delimiter', kind='RawStringDelimiterToken',
```
Comment (pschuh) [inline utils/gyb_syntax_support/ExprNodes.py]: Unrelated change?

File: test/Frontend/dump-parse.swift [hunk 2]
```diff
@@ -22,9 +22,9 @@ func bar() {
   // CHECK-NEXT:   (unresolved_decl_ref_expr type='{{[^']+}}' name=foo
   // CHECK-NEXT:   (unresolved_decl_ref_expr type='{{[^']+}}' name=foo
   // CHECK-AST: (brace_stmt
-  // CHECK-AST-NEXT:   (declref_expr type='{{[^']+}}' {{.*}} decl=main.(file).foo
-  // CHECK-AST-NEXT:   (declref_expr type='{{[^']+}}' {{.*}} decl=main.(file).foo
-  // CHECK-AST-NEXT:   (declref_expr type='{{[^']+}}' {{.*}} decl=main.(file).foo
+  // CHECK-AST-NEXT:   (declref_expr type='{{[^']+}}' {{.*}} decl=main.(file).foo{{.*}}
```
Comment (pschuh) [inline test/Frontend/dump-parse.swift]: Unrelated change?

File: lib/Sema/QuoteTransform.cpp [hunk 3]
```diff
@@ -0,0 +1,1361 @@
...
+
+  Expr *visitImplicitConversionExpr(ImplicitConversionExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    return makeQuote("Conversion", {quoteExpr(expr->getSubExpr()),
+                                    quoteType(expr->getType())});
+  }
+
+  Expr *visitInOutExpr(InOutExpr *expr) {
+    Breadcrumb bc(bcs, expr);
+    if (expr->isImplicit()) {
```
Comment (pschuh) [inline lib/Sema/QuoteTransform.cpp]: What is the general strategy around Exprs with isImplicit? Is this just a one-off case of doing something special in order to make things prettier? Or does some sort of correctness rely on being able to recover fully the original expr before things were rewritten?

**Review by pschuh — APPROVED** [review]

File: include/swift/AST/Decl.h [hunk 4]
```diff
@@ -6276,11 +6276,12 @@ AbstractStorageDecl::AccessorRecord::getAccessor(AccessorKind kind) const {
   }
   return nullptr;
 }
-  
+
 /// This represents a 'case' declaration in an 'enum', which may declare
 /// one or more individual comma-separated EnumElementDecls.
```
[the remark this answers is not in the document]
Comment (burmako) [reply include/swift/AST/Decl.h]: I think clang-format went a little overboard here. Reverted.

File: lib/AST/ConcreteDeclRef.cpp [hunk 5]
```diff
@@ -59,6 +60,9 @@ void ConcreteDeclRef::dump(raw_ostream &os) {
     getSubstitutions().dump(os, SubstitutionMap::DumpStyle::Minimal);
     os << ']';
   }
+
+  os << " USR=";
+  ide::printDeclUSR(getDecl(), os);
```
[the remark this answers is not in the document]
Comment (burmako) [reply lib/AST/ConcreteDeclRef.cpp]: This was useful to figure out USRs of references since these USRs are used as symbols in quoted trees. However this is not strictly necessary, so I reverted the change.

File: utils/gyb_syntax_support/ExprNodes.py (hunk 1 above)
Comment (burmako) [reply utils/gyb_syntax_support/ExprNodes.py]: I changed my editor to avoid trimming whitespace in C++ files to not break formatting in the Swift repo, but forgot to do the same for Python files. Reverted.

File: test/Frontend/dump-parse.swift (hunk 2 above)
Comment (burmako) [reply test/Frontend/dump-parse.swift]: This is caused by the USR printing change which has been reverted, so this change has been reverted as well.

File: lib/Sema/QuoteTransform.cpp (hunk 3 above)
Comment (burmako) [reply lib/Sema/QuoteTransform.cpp]: This is a one-off case because `InOutExpr` seems to be used both syntactically (to represent `&x`) and semantically (to represent certain references to in-out parameters). I've added a comment.

**Comment by burmako** [comment]
@swift-ci Please test tensorflow

**Comment by burmako** [comment]
@swift-ci please clean test tensorflow
